### PR TITLE
Support building swift/AST with MSVC on Windows

### PR DESCRIPTION
**Pull request by hughbe**
Relies on apple/swift-llvm#33 but can be merged before

### CODE REVIEW

File: include/swift/AST/Module.h [hunk 1]
```diff
@@ -744,7 +748,7 @@ class FileUnit : public DeclContext {
   // Only allow allocation of FileUnits using the allocator in ASTContext
   // or by doing a placement new.
   void *operator new(size_t Bytes, ASTContext &C,
-                     unsigned Alignment = alignof(FileUnit));
+                     unsigned Alignment = alignof(FileUnit&));
```
Comment (hughbe) [inline include/swift/AST/Module.h]: This fixes an MSVC error: `cannot instantiate abstract class FileUnit`. I'm not sure how correct this is (@jrose-apple you seem to know a lot of these sorts of things)

Comment (jrose-apple) [inline include/swift/AST/Module.h]: Definitely not correct, except coincidentally. Also confused; this operator occurs within the definition of FileUnit, so I would expect it to be able to calculate the alignment.

…well, okay, maybe not, that might require two passes. \*goes to check C++ standard\*

File: include/swift/AST/AnyFunctionRef.h [hunk 2]
```diff
@@ -151,7 +151,10 @@ class AnyFunctionRef {
     }
     llvm_unreachable("unexpected AnyFunctionRef representation");
   }
-  
+
+// Disable "only for use within the debugger" warning
+#pragma warning(push)
```
Comment (jrose-apple) [inline include/swift/AST/AnyFunctionRef.h]: *Clang* still has unknown pragma warnings on, so this should be guarded by a check for MSVC.

File: include/swift/AST/Decl.h [hunk 3]
```diff
@@ -819,7 +819,11 @@ class alignas(1 << DeclAlignInBits) Decl {
 
   // Make vanilla new/delete illegal for Decls.
   void *operator new(size_t Bytes) = delete;
+
+  // Work around MSVC error: attempting to reference a deleted function.
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: What does MSVC consider a reference that Clang doesn't?

File: include/swift/AST/Expr.h [hunk 4]
```diff
@@ -652,13 +652,27 @@ class TrailingCallArguments
     return *static_cast<const Derived *>(this);
   }
 
+  // Work around MSVC bug: can't infer llvm::trailing_objects_internal,
+  // even though we granted friend access to it.
   size_t numTrailingObjects(
+#if defined(_MSC_VER) && !defined(__clang__)
+      llvm::trailing_objects_internal::TrailingObjectsBase::OverloadToken<
```
Comment (jrose-apple) [inline include/swift/AST/Expr.h]: This seems really brittle. I still don't like it, but I don't have an alternative to play with right now.

File: include/swift/AST/LookupKinds.h [hunk 5]
```diff
@@ -65,8 +65,14 @@ enum NLOptions : unsigned {
   /// primary source file.
   ///
   /// \see NL_KnownDependencyMask
-  NL_KnownNoDependency =
+
+  // Work around MSVC bug: expression did not evaluate to a constant.
```
Comment (jrose-apple) [inline include/swift/AST/LookupKinds.h]: If you're going to do this, just write the constant unconditionally, and static_assert that it has the correct value.

File: lib/AST/ASTMangler.cpp [hunk 6]
```diff
@@ -36,11 +36,11 @@
 #include "llvm/Support/CommandLine.h"
 
 using namespace swift;
-using namespace NewMangling;
+using namespace swift::NewMangling;
 
 std::string NewMangling::mangleTypeForDebugger(Type Ty, const DeclContext *DC) {
   if (useNewMangling()) {
-    ASTMangler::ASTMangler NewMangler(/* DWARF */ true);
+    ASTMangler NewMangler(/* DWARF */ true);
```
Comment (jrose-apple) [inline lib/AST/ASTMangler.cpp]: Yeah, this just looks like a mistake.

File: lib/AST/NameLookup.cpp [hunk 7]
```diff
@@ -28,6 +28,11 @@
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/TinyPtrVector.h"
 
+#if defined(_MSC_VER) && !defined(__clang__)
```
Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: We can just include `<cassert>` unconditionally. What's `intrin.h` for?

Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: Oh, I see, it's the popcnt check below. I think it's fine to leave that out, since the static_assert will be checked in Clang anyway.

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/Module.h]: Okay, this actually *does* look correct—"When alignof is applied to a reference type, the result shall be the alignment of the referenced type." (C++11 [expr.alignof]p3)—but if MSVC is accepting one and not the other I'm a little concerned that it's doing the wrong thing. Can you test that it accepts this code?

```c++
struct Foo {
  bool b;
  static_assert(alignof(Foo&) == alignof(bool), "aligning reference rather than Foo");
};
```

Comment (jrose-apple) [inline include/swift/AST/Module.h]: Specific citation allowing this (which says nothing to whether MSVC implements it): "Within the class _member-specification_, the class is regarded as complete within function bodies, **default arguments,** and _brace-or-equal-initializers_ for non-static data members (including such things in nested classes)." (C++11 [class.mem]p2)

File: lib/AST/ASTMangler.cpp (hunk 6 above)
Comment (eeckstein) [inline lib/AST/ASTMangler.cpp]: yes, it was

File: include/swift/AST/Module.h (hunk 1 above)
Comment (hughbe) [inline include/swift/AST/Module.h]: The following modification of your example works:
```
struct Foo {
  bool b;
};

 static_assert(alignof(Foo&) == alignof(bool), "aligning reference rather than Foo");
```

Comment (hughbe) [inline include/swift/AST/Module.h]: ... your example gives me the errors
>source_file.cpp(12): error C2027: use of undefined type 'Foo'
>source_file.cpp(10): note: see declaration of 'Foo'
>source_file.cpp(12): error C2338: aligning reference rather than Foo

File: include/swift/AST/Expr.h (hunk 4 above)
Comment (hughbe) [inline include/swift/AST/Expr.h]: I hate this too, but seems to be the best I can figure out

File: include/swift/AST/Decl.h (hunk 3 above)
Comment (hughbe) [inline include/swift/AST/Decl.h]: I get the following error

> error C2280: 'void *swift::AST::ModuleDecl::__delDtor(unsigned int)': attempting to reference a deleted function 

The problem seems to be that MSVC generates this `__delDtor` intrinsic that I don't seem to have control over.

Comment (hughbe) [inline include/swift/AST/Decl.h]: Take a look at this link: https://msdn.microsoft.com/en-us/library/mt723604.aspx
There seems to be some mention of this error, but I'm not sure how applicable this is

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/Module.h]: Yeah, in that case it's not correct. We really do need to align based on the type and not on a reference to it.

File: include/swift/AST/Decl.h (hunk 3 above)
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Ah, maybe you would then have to delete `operator delete` in all the subclasses too.

It's possible marking it `virtual` would work?

File: lib/AST/NameLookup.cpp (hunk 7 above)
Comment (jrose-apple) [inline lib/AST/NameLookup.cpp]: Ah, both. :-) Please move `<cassert>` out of the `#if`, but don't bother asserting the popcount under MSVC (just `#if` it out).

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/Module.h]: Maybe we can do something silly like declaring a function `alignOfFileUnit` and then defining it at the bottom of the file as static inline.

Comment (hughbe) [inline include/swift/AST/Module.h]: Hmm, something like this:
```
class FileUnit {
public:
  // Only allow allocation of FileUnits using the allocator in ASTContext
  // or by doing a placement new.
  void *operator new(size_t Bytes, ASTContext &C,
                     unsigned Alignment = alignOfFileUnit());
};

static inline unsigned alignOfFileUnit() {
	return alignof(FileUnit);
}
```

doesn't work due to same error:

>1>C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(757): error C2259: 'swift::FileUnit': cannot instantiate abstract class (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(757): note: due to following members: (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(757): note: 'void swift::FileUnit::lookupValue(swift::ModuleDecl::AccessPathTy,swift::DeclName,swift::NLKind,llvm::SmallVectorImpl<swift::ValueDecl *> &) const': is abstract (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(556): note: see declaration of 'swift::FileUnit::lookupValue' (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(757): note: 'void swift::FileUnit::lookupObjCMethods(swift::ObjCSelector,llvm::SmallVectorImpl<swift::AbstractFunctionDecl *> &) const': is abstract (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(588): note: see declaration of 'swift::FileUnit::lookupObjCMethods' (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(757): note: 'swift::Identifier swift::FileUnit::getDiscriminatorForPrivateValue(const swift::ValueDecl *) const': is abstract (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)
1>  C:\Users\Lukey\Documents\GitHub\my-swift\swift\include\swift/AST/Module.h(631): note: see declaration of 'swift::FileUnit::getDiscriminatorForPrivateValue' (compiling source file C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTPrinter.cpp)

File: include/swift/AST/Decl.h (hunk 3 above)
Comment (hughbe) [inline include/swift/AST/Decl.h]: > Ah, maybe you would then have to delete operator delete in all the subclasses too.

This doesn't work - it just changes the error message from
> void swift::Decl::operator delete(void *)': attempting to reference a deleted function

to 
>void swift::ModuleDecl::operator delete(void *)': attempting to reference a deleted function

Making it virtual causes the error

> 'delete' cannot be a virtual function

File: include/swift/AST/LookupKinds.h (hunk 5 above)
Comment (hughbe) [inline include/swift/AST/LookupKinds.h]: I can't seem to reproduce this - I think it is a Visual Studio bug. I've removed the workaround

File: lib/AST/NameLookup.cpp (hunk 7 above)
Comment (hughbe) [inline lib/AST/NameLookup.cpp]: fixed, thanks

File: include/swift/AST/AnyFunctionRef.h (hunk 2 above)
Comment (hughbe) [inline include/swift/AST/AnyFunctionRef.h]: ah I assumed if this worked with clang-cl it would work with Clang, but I was wrong - fixed

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/Module.h]: Ah, that's a much more enlightening error.

Does the reference do the right thing *now?*

```c++
class Test {
  alignas(16) double x;
public:
  virtual void foo() = 0;
};

static_assert(alignof(Test&) == 16, "");
```

(Of course, maybe MSVC doesn't support overalignment, in which case the test won't work again.)

File: include/swift/AST/Decl.h (hunk 3 above)
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: All right, oh well. We can do the workaround.

File: include/swift/AST/Module.h (hunk 1 above)
Comment (jrose-apple) [inline include/swift/AST/Module.h]: Okay, then in that case we can use `alignof(FileUnit&)` in `alignOfFileUnit` and we should get the correct answer.

(I think this is actually important because DeclContexts are deliberately overaligned, and we rely on that to cram things in the low bits.)

**Comment by hughbe** [comment]
I've rebased and addressed your comments about the alignment issues

File: lib/AST/ASTContext.cpp [hunk 8]
```diff
@@ -276,8 +276,12 @@ struct ASTContext::Implementation {
     ~Arena() {
       for (auto &conformance : SpecializedConformances)
         conformance.~SpecializedProtocolConformance();
+      // Work around MSVC warning: local variable is initialized but
+      // not referenced.
+#pragma warning (disable: 4189)
```
Comment (jrose-apple) [inline lib/AST/ASTContext.cpp]: More unguarded pragmas. Also, I find it hard to believe that we'd get a warning on this code and not the previous line, so maybe this is no longer necessary?

File: lib/AST/ASTDumper.cpp [hunk 9]
```diff
@@ -1473,12 +1473,15 @@ void Stmt::print(raw_ostream &OS, unsigned Indent) const {
 // Printing for Expr and all subclasses.
 //===----------------------------------------------------------------------===//
 
-static raw_ostream &operator<<(raw_ostream &os, AccessSemantics accessKind) {
+// Ideally, we'd like to define this as an operator (<<), but MSVC reports many
+// ambiguous resolution errors, as it tries to overload according to the enum's
+// underlying type, not primarily the actual enum.
```
Comment (jrose-apple) [inline lib/AST/ASTDumper.cpp]: :-(

File: lib/AST/ASTMangler.cpp [hunk 10]
```diff
@@ -36,12 +36,12 @@
...
 using namespace swift;
-using namespace NewMangling;
+using namespace swift::NewMangling;
 
 std::string NewMangling::mangleTypeForDebugger(Type Ty, const DeclContext *DC) {
   if (useNewMangling()) {
-    ASTMangler::ASTMangler NewMangler(/* DWARF */ true);
-    return NewMangler.mangleTypeForDebugger(Ty, DC);
+    ASTMangler NewMangler(/* DWARF */ true);
+    return NewMangler.mangleType(Ty, DC);
```
Comment (jrose-apple) [inline lib/AST/ASTMangler.cpp]: Why the change here? These do different things. cc @eeckstein

**Review by jrose-apple — COMMENTED** [review]
Just a few last comments.

File: lib/AST/ASTMangler.cpp (hunk 10 above)
Comment (hughbe) [inline lib/AST/ASTMangler.cpp]: whoops - merge conflict

File: lib/AST/ASTContext.cpp (hunk 8 above)
Comment (hughbe) [inline lib/AST/ASTContext.cpp]: We do get the following warning:

> C:\Users\Lukey\Documents\GitHub\my-swift\swift\lib\AST\ASTContext.cpp(282): warning C4189: 'conformance': local variable is initialized but not referenced

very weird...

File: lib/AST/ASTDumper.cpp (hunk 9 above)
Comment (hughbe) [inline lib/AST/ASTDumper.cpp]: Okay, I have a much better fix that suddenly came to me when you commented. Basically, the problem was that the code referring to this `<<` operator is in an anonymous namespace. Moving `<<` into that anonymous namespace fixes this. Unfortunately, MSVC wasn't very helpful as it reported the error as an ambiguous overload error for `unsigned char`

File: lib/AST/ASTDumper.cpp [hunk 11]
```diff
@@ -1473,17 +1473,17 @@ void Stmt::print(raw_ostream &OS, unsigned Indent) const {
...
+namespace {
 static raw_ostream &operator<<(raw_ostream &os, AccessSemantics accessKind) {
-  switch (accessKind) {
-  case AccessSemantics::Ordinary: return os;
-  case AccessSemantics::DirectToStorage: return os << " direct_to_storage";
-  case AccessSemantics::DirectToAccessor: return os << " direct_to_accessor";
-  case AccessSemantics::BehaviorInitialization: return os << " behavior_init";
-  }
-  llvm_unreachable("bad access kind");
+	switch (accessKind) {
```
Comment (jrose-apple) [inline lib/AST/ASTDumper.cpp]: Tabs snuck in here.

Comment (hughbe) [inline lib/AST/ASTDumper.cpp]: Thanks - this should be fixed now

File: include/swift/AST/Decl.h (hunk 3 above)
Comment (jckarter) [inline include/swift/AST/Decl.h]: This looks like an MSVC bug. I presume the deleting destructor is part of its ABI, but that entry point seems like it ought to be transitively `delete`-d if `operator delete` is deleted.

Comment (jrose-apple) [inline include/swift/AST/Decl.h]: I can believe it's an MSVC bug, but I don't mind us working around it for now.

Comment (hughbe) [inline include/swift/AST/Decl.h]: I've submitted a connect bug for this, at https://connect.microsoft.com/VisualStudio/feedback/details/3116505

**Comment by hughbe** [comment]
I've made a connect bug for the numTrailingObjects problem: https://connect.microsoft.com/VisualStudio/feedback/details/3116517

If anyone's curious, they can try coming up with a better workaround by compiling MSVC code online here: http://rextester.com/XIJ43194

File: include/swift/AST/AnyFunctionRef.h [hunk 12]
```diff
@@ -151,7 +151,12 @@ class AnyFunctionRef {
     }
     llvm_unreachable("unexpected AnyFunctionRef representation");
   }
-  
+
+// Disable "only for use within the debugger" warning.
```
Comment (hughbe) [inline include/swift/AST/AnyFunctionRef.h]: https://connect.microsoft.com/VisualStudio/feedback/details/3116508/msvc-shouldnt-show-errors-using-deprecated-methods-from-deprecated-methods

File: include/swift/AST/Types.h [hunk 13]
```diff
@@ -1352,7 +1352,7 @@ class ParameterTypeFlags {
     NumBits = 3
   };
   OptionSet<ParameterFlags> value;
-  static_assert(NumBits < 8*sizeof(value), "overflowed");
+  static_assert(NumBits < 8*sizeof(OptionSet<ParameterFlags>), "overflowed");
```
Comment (hughbe) [inline include/swift/AST/Types.h]: Dupe of https://connect.microsoft.com/VisualStudio/feedback/details/3116640/msvc-doesnt-support-sizeof-value-in-a-static-assert-in-a-class

File: include/swift/AST/Module.h (hunk 1 above)
Comment (hughbe) [inline include/swift/AST/Module.h]: https://connect.microsoft.com/VisualStudio/feedback/details/3116815

File: lib/AST/ASTContext.cpp [hunk 14]
```diff
@@ -275,8 +275,16 @@ struct ASTContext::Implementation {
     ~Arena() {
       for (auto &conformance : SpecializedConformances)
         conformance.~SpecializedProtocolConformance();
+      // Work around MSVC warning: local variable is initialized but
+      // not referenced.
+#if defined(_MSC_VER)
+#pragma warning (disable: 4189)
```
Comment (hughbe) [inline lib/AST/ASTContext.cpp]: This seems to be a bug in MSVC: I've been trying hard to reproduce this came, but can't.

File: lib/AST/ASTDumper.cpp [hunk 15]
```diff
@@ -1473,6 +1473,7 @@ void Stmt::print(raw_ostream &OS, unsigned Indent) const {
 // Printing for Expr and all subclasses.
 //===----------------------------------------------------------------------===//
 
+namespace {
```
Comment (hughbe) [inline lib/AST/ASTDumper.cpp]: Again, this looks to be an MSVC bug, but I can't reproduce it :(

**Comment by hughbe** [comment]
@slavapestov PTAL. I think the only thing holding back this one getting merged is the gross hack for https://connect.microsoft.com/VisualStudio/feedback/details/3116517. I'd like to figure out a work around that doesn't make us diverge from upstream LLVM.

If anyone has some spare time, check out a simple reproduction here: http://rextester.com/XIJ43194. This one also lets you compile C++ online if you don't have a Windows PC

**Comment by slavapestov** [comment]
@swift-ci Please smoke test
